ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java

```diff
@@ -168,51 +171,80 @@ private int sizeExcludingBlobAndCrcSize() {
     return sizeExcludingBlobAndCrc;
   }
 
+  private void prepareBuffer() {
+    if (bufferToSend == null) {
+      // this is the first time this method was called, prepare the buffer to send the header and other metadata
+      // (everything except the blob content).
+      bufferToSend = ByteBuffer.allocate(sizeExcludingBlobAndCrcSize());
+      writeHeader();
+      int crcStart = bufferToSend.position();
+      bufferToSend.put(blobId.toBytes());
+      BlobPropertiesSerDe.serializeBlobProperties(bufferToSend, properties);
```

Comment:
Is it possible to convert BlobProperties and userMeatadata to bytebuf first, and then use `Bytebuf bufferToSend`?

Reply:
Use bytebuf as usermetadata's data type is currently in my plan, but I would do blob data first since blob data is usually much larger than the user metadata.

Reply:
Yes, but if you can make the change in one shot, you can eliminate ByteBuffer bufferToSend. Everything can be ByteBuf based in PutRequest.

Reply:
The problem is bufferToSend belongs to the it's parent class RequestOrResponse. To change the type of bufferToSend would trigger lots of changes for other classes as well.

Reply:
Any plan to change it?

Reply:
We can change it later. I have a jira to change user metadata to use netty bytebuf. will do it with this jira.